File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/midi/viewmodel/MIDIViewModel.kt
package com.musicmuni.voxatrace.demo.sections.midi.viewmodel

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.musicmuni.voxatrace.sonix.SonixMidiSynthesizer
import com.musicmuni.voxatrace.sonix.SonixPlayer
import com.musicmuni.voxatrace.sonix.midi.MidiNote
import io.github.aakira.napier.Napier
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream

/**
 * ViewModel for MIDI synthesis using SonixMidiSynthesizer.
 */
class MIDIViewModel : ViewModel() {

    // MARK: - Published State

    private val _status = MutableStateFlow("Ready")
    val status: StateFlow<String> = _status.asStateFlow()

    private val _isSynthesizing = MutableStateFlow(false)
    val isSynthesizing: StateFlow<Boolean> = _isSynthesizing.asStateFlow()

    private val _outputFile = MutableStateFlow<File?>(null)
    val outputFile: StateFlow<File?> = _outputFile.asStateFlow()

    private val _synthesizerVersion = MutableStateFlow<String?>(null)
    val synthesizerVersion: StateFlow<String?> = _synthesizerVersion.asStateFlow()

    private val _selectedSampleRate = MutableStateFlow(44100)
    val selectedSampleRate: StateFlow<Int> = _selectedSampleRate.asStateFlow()

    private val _isPlaying = MutableStateFlow(false)
    val isPlaying: StateFlow<Boolean> = _isPlaying.asStateFlow()

    // MARK: - Private State

    private var player: SonixPlayer? = null
    private val observerJobs = mutableListOf<Job>()

    // MARK: - Actions

    fun setSelectedSampleRate(rate: Int) {
        if (!_isSynthesizing.value) {
            _selectedSampleRate.value = rate
        }
    }

    fun synthesize(context: Context) {
        viewModelScope.launch {
            _isSynthesizing.value = true
            _status.value = "Synthesizing (${_selectedSampleRate.value / 1000}kHz)..."

            try {
                // Copy soundfont to file
                val sfFile = withContext(Dispatchers.IO) {
                    copyAssetToFile(context, "harmonium.sf2")
                }

                // C major scale (timing in milliseconds)
                val notes = listOf(
                    MidiNote(note = 60, startTime = 0f, endTime = 400f),     // C4
                    MidiNote(note = 62, startTime = 500f, endTime = 900f),   // D4
                    MidiNote(note = 64, startTime = 1000f, endTime = 1400f), // E4
                    MidiNote(note = 65, startTime = 1500f, endTime = 1900f), // F4
                    MidiNote(note = 67, startTime = 2000f, endTime = 2400f), // G4
                    MidiNote(note = 69, startTime = 2500f, endTime = 2900f), // A4
                    MidiNote(note = 71, startTime = 3000f, endTime = 3400f), // B4
                    MidiNote(note = 72, startTime = 3500f, endTime = 4400f), // C5
                )

                val outFile = File(context.cacheDir, "midi_output.wav")

                val (success, version) = withContext(Dispatchers.IO) {
                    val synth = SonixMidiSynthesizer.Builder()
                        .soundFontPath(sfFile.absolutePath)
                        .sampleRate(_selectedSampleRate.value)
                        .onError { error ->
                            Napier.e("Synth error: $error")
                        }
                        .build()

                    val result = synth.synthesizeFromNotes(
                        notes = notes,
                        outputPath = outFile.absolutePath
                    )
                    Pair(result, synth.version)
                }

                if (!isActive) return@launch

                _synthesizerVersion.value = version
                if (success && outFile.exists() && outFile.length() > 0) {
                    _outputFile.value = outFile
                    _status.value = "Generated (${_selectedSampleRate.value / 1000}kHz): ${outFile.length() / 1024} KB"

                    // Load into SonixPlayer for playback
                    player?.release()
                    val newPlayer = SonixPlayer.create(outFile.absolutePath)
                    player = newPlayer

                    // Setup observer
                    observerJobs.add(viewModelScope.launch {
                        newPlayer.isPlaying.collect { playing ->
                            _isPlaying.value = playing
                        }
                    })
                } else {
                    _status.value = "Synthesis failed"
                }
            } catch (e: Exception) {
                Napier.e("MIDI synthesis failed", e)
                if (isActive) _status.value = "Error: ${e.message}"
            } finally {
                if (isActive) _isSynthesizing.value = false
            }
        }
    }

    fun play() {
        player?.play()
    }

    fun stop() {
        player?.stop()
    }

    override fun onCleared() {
        super.onCleared()
        observerJobs.forEach { it.cancel() }
        player?.release()
    }

    private fun copyAssetToFile(context: Context, assetName: String): File {
        val file = File(context.cacheDir, assetName)
        if (!file.exists()) {
            context.assets.open(assetName).use { input ->
                FileOutputStream(file).use { output ->
                    input.copyTo(output)
                }
            }
        }
        return file
    }
}

File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/MIDI/MIDIViewModel.swift
import Foundation
import Combine
import VoxaTrace

/// ViewModel for MIDI synthesis using SonixMidiSynthesizer.Builder.
///
/// ## VoxaTrace Integration (~15 lines)
/// ```swift
/// // 1. Create synthesizer with Builder pattern
/// let synth = SonixMidiSynthesizer.Builder()
///     .soundFontPath(path: sfPath)
///     .sampleRate(rate: sampleRate)
///     .build()
///
/// // 2. Synthesize
/// let success = synth.synthesizeFromNotes(notes: notes, outputPath: outputPath)
///
/// // 3. Play result with SonixPlayer
/// player = try await SonixPlayer.create(source: outputPath, audioSession: .playback)
/// ```
@MainActor
final class MIDIViewModel: ObservableObject {

    // MARK: - Published State

    @Published private(set) var status = "Ready"
    @Published private(set) var isSynthesizing = false
    @Published private(set) var isPlaying = false
    @Published private(set) var synthesizerVersion: String?
    @Published var selectedSampleRate: Int32 = 44100

    // MARK: - Private

    private var outputFile: URL?
    private var player: SonixPlayer?
    private var playingTask: Task<Void, Never>?

    let sampleRates: [Int32] = [44100, 48000]

    // MARK: - Lifecycle

    func onDisappear() {
        playingTask?.cancel()
        player?.release()
    }

    // MARK: - Actions

    func synthesize() {
        Task {
            isSynthesizing = true
            status = "Synthesizing (\(selectedSampleRate / 1000)kHz)..."

            do {
                guard let sfPath = copyAssetToFile(name: "harmonium", ext: "sf2") else {
                    status = "Soundfont not found"
                    isSynthesizing = false
                    return
                }

                // C major scale (timing in milliseconds)
                let notes: [MidiNote] = [
                    MidiNote(note: 60, startTime: 0, endTime: 400),     // C4
                    MidiNote(note: 62, startTime: 500, endTime: 900),   // D4
                    MidiNote(note: 64, startTime: 1000, endTime: 1400), // E4
                    MidiNote(note: 65, startTime: 1500, endTime: 1900), // F4
                    MidiNote(note: 67, startTime: 2000, endTime: 2400), // G4
                    MidiNote(note: 69, startTime: 2500, endTime: 2900), // A4
                    MidiNote(note: 71, startTime: 3000, endTime: 3400), // B4
                    MidiNote(note: 72, startTime: 3500, endTime: 4400), // C5
                ]

                let tempDir = FileManager.default.temporaryDirectory
                let outFile = tempDir.appendingPathComponent("midi_output.wav")

                // Builder API: Advanced configuration
                let synth = SonixMidiSynthesizer.Builder()
                    .soundFontPath(path: sfPath)
                    .sampleRate(rate: selectedSampleRate)
                    .build()

                synthesizerVersion = synth.version

                let success = synth.synthesizeFromNotes(
                    notes: notes,
                    outputPath: outFile.path
                )

                if success {
                    let attrs = try FileManager.default.attributesOfItem(atPath: outFile.path)
                    let fileSize = (attrs[.size] as? Int64) ?? 0
                    outputFile = outFile
                    status = "Generated (\(selectedSampleRate / 1000)kHz): \(fileSize / 1024) KB"

                    await loadPlayer(path: outFile.path)
                } else {
                    status = "Synthesis failed"
                }
            } catch {
                status = "Error: \(error.localizedDescription)"
            }

            isSynthesizing = false
        }
    }

    func play() {
        player?.play()
    }

    func stop() {
        player?.stop()
    }

    // MARK: - Computed Properties

    var canPlay: Bool {
        outputFile != nil && !isSynthesizing && !isPlaying
    }

    // MARK: - Private Methods

    private func loadPlayer(path: String) async {
        playingTask?.cancel()
        player?.release()

        do {
            let newPlayer = try await SonixPlayer.create(source: path, audioSession: .playback)
            player = newPlayer
            playingTask = newPlayer.observeIsPlaying { [weak self] playing in
                self?.isPlaying = playing
            }
        } catch {
            status = "Player error: \(error.localizedDescription)"
        }
    }
}
